packages/ses/src/module-instance.js

```diff
@@ -0,0 +1,335 @@
+import { performEval } from './evaluate.js';
+import { getCurrentRealmRec } from './realm-rec.js';
+
+const { create, entries, keys, freeze, seal, defineProperty: defProp } = Object;
+
+// q, for enquoting strings in error messages.
+const q = JSON.stringify;
+
+// `makeModuleInstance` takes a module's analysis (the private data of a module
+// static record), the live import namespace, and a global object, and produces
+// a module instance.
+// The module analysis is a subset of a module's compartment record.
+// The module instance carries the module exports namespace (the "module"),
+// notifiers to update the module's internal import namespace, and an
+// idempotent execute function.
+// The actual module namespace is a proxy that will delegate to the internal
+// module import namespace before the module executes.
+export const makeModuleInstance = (
+  moduleAnalysis,
+  importedInstances,
+  globalObject,
+) => {
+  const {
+    functorSource,
+    fixedExportMap,
+    liveExportMap,
+    exportAlls,
+  } = moduleAnalysis;
+
+  // {_exportName_: getter} module exports namespace object
+  const module = create(null);
```

Comment:
There are too many things referred to as "module" without modifier. This is already a problem when speaking as well. I understand that this isn't the moduleNS object itself, because that will be the proxy. But we still need something here. This is the "module <what?>"

I should have mentioned in the last PR that I felt likewise, though not a strongly, about "moduleRecord" for "module compartment record". OTOH, "moduleCompartmentRecord" is too horribly long. I don't have a suggestion for that one either.

Naming is hard.

Reply:
I felt the same tension in the prior PR and used context to afford some brevity. In the documentation I’m drafting, I’m going so far as to spell out “internal module exports namespace”, “module exports namespace publicly visible proxy”, “module compartment static record”, “module compartment instance dynamic record”, and parenthesizing the brief form for local variables where it’s unambiguous in context. The problem is manifold larger to capture all of the internal concepts.

The externally visible concepts should be the most pithy. We can afford verbosity internally, or buy some terseness with context, at some risk of confusion. For example, the “module analysis” is an internal concept that corresponds to “module static record”, but is presently called “module static record” in transform-module. That’s a wrinkle I’d like to iron out in a future change.

If we bless one thing to have the synecdoche “module”, I think we agree it would be “module exports namespace (the publicly visible proxy)”, which buys us the short name `compartment.module(fullSpec)`. But for clarity with private names, we can still be fully explicit, with suitably long names.

Naming is indeed hard. I will land this as-is only because we agree upon the need to reserve the right to continue ironing out these internal names.